### ISSUE TITLE
sessions REST API failed with non-root realm

### ISSUE THREAD

**Issue by ftntming** [OpenIdentityPlatform/OpenAM#305]
**Describe the bug**
Can use REST API to login user to non-root realm. The returned tokenId is valid.
However, the API to verify the token returns 403
**This does not happen to root realm and amAdmin user. Only happen to non-root realm and LDAP auth user.**

**To Reproduce**
 user login to openam with non-root realm, e.g. realm2,  using API, the tokenId will be returned correctly, e.g.
```
curl --location --request POST 'http://openam-0.openam.fvinfra.svc.cluster.local:8080/openam/json/authenticate?realm=realm2&module=my_ldap&authIndexType=module&authIndexValue=my_ldap' \
--header 'X-OpenAM-Username: user88' \
--header 'X-OpenAM-Password: USER88@pass'
```
Return token is 
```
{"tokenId":"AQIC5wM2LY4SfcwSipioB16IAPR4mqB5YyjKwWUpnynpF0s.*AAJTSQACMDEAAlNLABM3MzAyNzczMjg5Mjc4MDE5NTYxAAJTMQAA*","successUrl":"/openam/console","realm":"/realm2"}
```
However, when I try to verify the token using 
```
curl --location --request POST 'http://openam-0.openam.fvinfra.svc.cluster.local:8080/openam/json/sessions?_action=getSessionInfo&realm=realm2' --header 'iPlanetDirectoryPro: AQIC5wM2LY4SfcwSipioB16IAPR4mqB5YyjKwWUpnynpF0s.*AAJTSQACMDEAAlNLABM3MzAyNzczMjg5Mjc4MDE5NTYxAAJTMQAA*'
```
I got: 
```
{"code":403,"reason":"Forbidden","message":"Forbidden"}
```
**Expected behavior**
API should return session/tokenId info

**Screenshots**
N/A

**Desktop (please complete the following information):**
N/A. curl command works in terminal

**Smartphone (please complete the following information):**
N/A

**Additional context**
**This does not happen to root realm and amAdmin user. Only happen to non-root realm and LDAP auth user.**
```
curl --location --request POST 'http://openam-0.openam.fvinfra.svc.cluster.local:8080/openam/json/authenticate' --header 'X-OpenAM-Username: amAdmin' --header 'X-OpenAM-Password: 12-password@123'

{"tokenId":"AQIC5wM2LY4SfcwClBvigOt46ff1D73EuiHtx9BheFHYzHs.*AAJTSQACMDEAAlNLABM1MjE1MTY4ODkzMjk2MTA2OTgyAAJTMQAA*","successUrl":"/openam/console","realm":"/"}
```
Verify "amAdmin" tokenID:
```
curl --location --request POST 'http://openam-0.openam.fvinfra.svc.cluster.local:8080/openam/json/realms/root/sessions/?_action=getSessionInfo' --header 'iPlanetDirectoryPro: AQIC5wM2LY4SfcwClBvigOt46ff1D73EuiHtx9BheFHYzHs.*AAJTSQACMDEAAlNLABM1MjE1MTY4ODkzMjk2MTA2OTgyAAJTMQAA*'


{"username":"amAdmin","universalId":"id=amAdmin,ou=user,dc=example,dc=com","realm":"/","sessionHandle":"shandle:AQIC5wM2LY4SfcxSBO1CO9suiFWAhcYQJ5wmcHziL4t5pUY.*AAJTSQACMDEAAlNLABM1MjE1MTY4ODkzMjk2MTA2OTgyAAJTMQAA*","latestAccessTime":"2020-09-02T08:24:53Z","maxIdleExpirationTime":"2020-09-02T08:54:53Z","maxSessionExpirationTime":"2020-09-02T10:20:52Z"}

```

**Comment by dairoca90** [comment]
There is a workaround for this?

**Comment by dairoca90** [comment]
It seems that the problem is the request, you can send it has the example and should work
`curl \
--request POST \
--Cookie "iPlanetDirectoryPro=AQIC5wM2LY4SfcyN7.....AAJTMQAA*" \
--header "Accept-API-Version: protocol=1.0,resource=2.0" \
--header "Content-Type: application/json" \
"http://am.openam.com/openam/json/sessions?_action=getSessionInfo&tokenId=AQIC5wM2LY4SfcyN7.....AAJTMQAA*"`

**Comment by sp193** [comment]
This doesn't seem to be correct. It is working in ForgeRock AM 5.5.1, which is what I was using prior to trying out this flavour of OpenAM.

This API is supposed to use the session token's ID from the request, or (optionally) the tokenId from the "tokenId" parameter:
As seen [here](https://github.com/OpenIdentityPlatform/OpenAM/blob/master/openam-core-rest/src/main/java/org/forgerock/openam/core/rest/session/TokenOwnerAuthzModule.java#L140
):
```
    boolean isTokenOwner(Context context, Request request) throws ResourceException, SSOException {
        String loggedInUserId = getUserId(context);

        String tokenId =  request.getResourcePath(); // infer from the token from resource path
        if (StringUtils.isEmpty(tokenId)) {
            //if there's no tokenId then is it supplied as additional parameter
            tokenId = request.getAdditionalParameter(tokenIdParameter);
        }

        final String queryUsername = ssoTokenManager.createSSOToken(tokenId).getProperty(Constants.UNIVERSAL_IDENTIFIER);
        return queryUsername.equalsIgnoreCase(loggedInUserId);
    }
```

At this point, there seems to be no tokenId captured and passed via the resource path of ActionRequest (the request parameter). Since there is no tokenId (it is ""), the authorization eventually fails. The context does recognize the user's DN, so it knows about the logged-in user.

I haven't been able to figure out where it's going wrong. I don't know how the resource path is set.

It works for amAdmin because of the AdminOnlyAuthzModule passing amAdmin.

This makes OpenAM incompatible with new versions of Forgerock's IG, when it tries to validate the session.

**Comment by sp193** [comment]
I am sorry, but it looks like it could be working correctly... for API resource 2.0. Or just correct as of when this version of OpenAM was forked.
This confusion probably came about because we're working on an OpenAM that deviated from ForgeRock's AM at some point in history.

I will write about what was found, in case this helps somebody else.

The place where the ActionRequest is "generated" is at org.forgerock.json.resource.Router.handleAction(). Here, the resourcePath may be set to the next segment of the URL path, for which there is nothing (if we called "/sessions", it is the last segment in the path). The resourcePath becomes the tokenId for TokenOwnerAuthzModule.isTokenOwner(), but only if we're talking about the structure of API resource 1.2.

In the API docs for AM 5.5.1, there are 3 sets of /sessions APIs. Their syntax are roughly something like:
1.2: /sessions/<tokenId>?_action=validate (note: there is no getSessionInfo at 1.2)
2.0: /sessions?tokenId=<tokenId>
3.0: /sessions

3.0 was when the tokenId may be specified inside the body. It might be also when ForgeRock made it possible for AM to use the session token as input. [AM 5.5 was also when the 3.0 resource for /sessions was added](https://backstage.forgerock.com/docs/am/5.5/release-notes/).

For comparison: I decompiled this class in AM 5.5.1 and found that the implementation of TokenOwnerAuthzModule.isTokenOwner is different. Here, getTokenID() returns the token, which may be from the cookie.
```
  boolean isTokenOwner(Context context, Request request) throws ResourceException, SSOException {
    String loggedInUserId = getUserId(context);
    String tokenId = SessionResourceUtil.getTokenID(context, request);
    String queryUsername = this.ssoTokenManager.createSSOToken(tokenId).getProperty("sun.am.UniversalIdentifier");
    return queryUsername.equalsIgnoreCase(loggedInUserId);
  }
```

And in SessionResourceUtil.getTokenID:
```
  public static String getTokenID(Context context, Request request) {
    String cookieName = null;
    String tokenId = getTokenIdFromParamOrBody(request);
    if (tokenId == null) {
      cookieName = SystemProperties.get("com.iplanet.am.cookie.name", "iPlanetDirectoryPro");
      tokenId = getTokenIdFromHeader(context, cookieName);
    } 
    if (tokenId == null)
      tokenId = getTokenIdFromCookie(context, cookieName); 
    return tokenId;
  }
```

Therefore, I think this issue can be closed because it's not really a bug.

But on the other hand: In AM 5.1's documentation, it [is suggested that the tokenId parameter was optional](https://backstage.forgerock.com/docs/am/5.1/authentication-guide/index.html#rest-api-session-information). So it might be a design issue that was fixed.

EDIT: seems to be also in the [5.0 documentation](https://backstage.forgerock.com/docs/am/5/authentication-guide/index.html#rest-api-session-information). So should we be using the same logic from AM?
